name-as/src/nma.rs

```diff
@@ -312,35 +452,101 @@ fn assemble_i(i_struct: &mut I, i_args: Vec<&str>) -> Result<u32, &'static str>
     Ok(result)
 }
 
+/// Assembles a J-type instruction
+fn assemble_j(
+    j_struct: &mut J,
+    j_args: Vec<&str>,
+    labels: &mut HashMap<&str, u32>,
+) -> Result<u32, &'static str> {
+    enforce_length(&j_args, 2)?;
+
+    let jump_address: u32 = labels[j_args[1]];
+    println!("Masking jump address");
+    println!("Jump address original: {}", jump_address);
+    let mut masked_jump_address = mask_u32(jump_address, 26);
+    println!("Jump address masked: {}", masked_jump_address);
+    if jump_address != masked_jump_address {
+        panic!("Tried to assemble illegal jump address");
+    }
+
+    // Byte-align jump address
+    masked_jump_address >>= 2;
+
+    let mut opcode = j_struct.opcode;
+
+    // Mask
+    println!("Masking opcode");
+    opcode = mask_u8(opcode, 6);
+    // No need to mask imm, it's already a u16
+
+    // opcode : 31 - 26
+    let mut result: u32 = opcode.into();
+
+    // imm :    25 - 0
+    println!("imm: {}", masked_jump_address);
+    result = (result << 26) | masked_jump_address;
+
+    println!(
+        "0x{:0shortwidth$x} {:0width$b}",
+        result,
+        result,
+        shortwidth = 8,
+        width = 32
+    );
+    Ok(result)
+}
+
 // General assembler entrypoint
 pub fn assemble(args: &Args) -> Result<(), &'static str> {
+    // IO Setup
     let input_fn = &args.input_as;
     let output_fn = &args.output_as;
 
+    let output_file: File = match File::create(output_fn) {
+        Ok(v) => v,
+        Err(_) => return Err("Failed to open output file"),
+    };
+
+    // Read input
     let file_contents: String = match fs::read_to_string(input_fn) {
         Ok(v) => v,
         Err(_) => return Err("Failed to read input file contents"),
     };
 
-    let mut tokens = tokenize(&file_contents);
+    // Parse labels (TODO : Make more robust later)
+    let mut current_addr: u32 = MIPS_TEXT_ADDRESS;
+    let mut labels: HashMap<&str, u32> = HashMap::new();
+    for line in file_contents.lines() {
+        if line.is_empty() {
+            continue;
+        }
 
-    let output_file: File = match File::create(output_fn) {
-        Ok(v) => v,
-        Err(_) => return Err("Failed to open output file"),
-    };
+        if let Ok(label_str) = label(line.trim()) {
+            println!("Inserting label {} at {:x}", label_str, current_addr);
+            labels.insert(label_str, current_addr);
+            continue;
+        }
 
+        current_addr += MIPS_INSTR_BYTE_WIDTH;
+    }
+    current_addr = MIPS_TEXT_ADDRESS;
+
+    // Tokenize input
+    let mut tokens = tokenize(&file_contents);
+
+    // Assembler setup
     let mut state = AssemblerState::Initial;
     let mut r_struct: R = R {
         shamt: 0,
         funct: 0,
         form: RForm::None,
     };
-    let mut r_args: Vec<&str> = Vec::new();
     let mut i_struct: I = I {
         opcode: 0,
         form: IForm::None,
     };
-    let mut i_args: Vec<&str> = Vec::new();
+    let mut j_struct: J = J { opcode: 0 };
```

Comment:
Didn't catch this earlier but you don't need to type annotate these. Just write:

```rust
let mut j_struct = J {opcode : 0};
```